tiled/catalog/adapter.py

```diff
@@ -574,13 +581,45 @@ async def create_node(
                         init_storage, data_uri, data_source.structure
                     )
                     data_source.assets.extend(assets)
-                data_source_orm = orm.DataSource(
-                    structure=_prepare_structure(
+                if data_source.structure is None:
+                    structure_id = None
+                else:
+                    # Obtain and hash the canonical (RFC 8785) representation of
+                    # the JSON structure.
+                    structure = _prepare_structure(
                         structure_family, data_source.structure
-                    ),
+                    )
+                    structure_id = compute_structure_id(structure)
+                    # The only way to do "insert if does not exist" i.e. ON CONFLICT
+                    # is to invoke dialect-specific insert.
```

Comment:
Can you avoid the dialect-specific clause by doing the search first? You're in the same transaction.

Alternatively, if I'm reading correctly, can you import the `insert` within the dialect select condition, then do the actual work in one piece of code?


```
if self.context.engine.dialect.name == "sqlite":
  from sqlalchemy.dialects.sqlite import insert

elif self.context.engine.dialect.name == "postgresql":
   from sqlalchemy.dialects.postgresql import insert

else:
   assert FALSE

statement = (insert(orm.Structure).values(
    id=structure_id,
    structure=structure,
    )
).on_conflict_do_nothing(index_elements=["id"])
await db.execute(statement)

```

Reply:
Good call. I refactored for now. It's possible we can remove the dialect-specific clause too. I need to run some experiments to check my understanding of transactions in SQLAlchemy and how they map onto the `SAVEPOINT` concept in PG and SQlite. Leaving that for a follow-up.